metrics/metrics.go

```diff
@@ -0,0 +1,115 @@
+package metrics
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/lucas-clemente/quic-go/logging"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+// Measures
+var (
+	connections = stats.Int64("connections", "number of QUIC connections", stats.UnitDimensionless)
+)
+
+// Tags
+var (
+	keyPerspective, _ = tag.NewKey("perspective")
+	keyIPVersion, _   = tag.NewKey("ip_version")
+)
+
+// Views
+var (
+	ConnectionsView = &view.View{
+		Measure:     connections,
+		TagKeys:     []tag.Key{keyPerspective, keyIPVersion},
+		Aggregation: view.Count(),
+	}
+)
+
+// DefaultViews collects all OpenCensus views for metric gathering purposes
+var DefaultViews = []*view.View{
+	ConnectionsView,
+}
+
+type tracer struct{}
+
+var _ logging.Tracer = &tracer{}
+
+// NewTracer creates a new metrics tracer.
+func NewTracer() logging.Tracer { return &tracer{} }
+
+func (t *tracer) TracerForServer(logging.ConnectionID) logging.ConnectionTracer {
+	return newConnTracer(t, logging.PerspectiveServer)
+}
+
+func (t *tracer) TracerForClient(logging.ConnectionID) logging.ConnectionTracer {
+	return newConnTracer(t, logging.PerspectiveClient)
+}
+
+type connTracer struct {
+	perspective logging.Perspective
+	tracer      logging.Tracer
+}
+
+func newConnTracer(tracer logging.Tracer, perspective logging.Perspective) logging.ConnectionTracer {
+	return &connTracer{
+		perspective: perspective,
+		tracer:      tracer,
+	}
+}
+
+var _ logging.ConnectionTracer = &connTracer{}
+
+func (t *connTracer) StartedConnection(local, _ net.Addr, _ logging.VersionNumber, _, _ logging.ConnectionID) {
+	perspectiveTag := tag.Upsert(keyPerspective, perspective(t.perspective).String())
+
+	var ipVersionTag tag.Mutator
+	if udpAddr, ok := local.(*net.UDPAddr); ok {
+		// If ip is not an IPv4 address, To4 returns nil.
+		// Note that there might be some corner cases, where this is not correct.
+		// See https://stackoverflow.com/questions/22751035/golang-distinguish-ipv4-ipv6.
+		if udpAddr.IP.To4() == nil {
+			ipVersionTag = tag.Upsert(keyIPVersion, "IPv6")
+		} else {
+			ipVersionTag = tag.Upsert(keyIPVersion, "IPv4")
+		}
+	} else {
+		ipVersionTag = tag.Upsert(keyIPVersion, "unknown")
+	}
+
+	stats.RecordWithTags(
+		context.Background(),
+		[]tag.Mutator{perspectiveTag, ipVersionTag},
+		connections.M(1),
+	)
+}
+
+func (t *connTracer) ClosedConnection(logging.CloseReason)                     {}
```

Comment:
I know that opencensus doesn't technically have the concept of a gauge but I am pretty sure that you can just do a `connection.M(-1)`. So we should be able to model `activeConnections` and not just the total number of connections that have been opened since the program started. There may be some issues with this approach if the metrics collector restarts and the process doesn't but I don't think so.

Reply:
I'm a bit confused now. I had thought of this counter to be plotted as something like `rate(connections[30s])`, so it would measure the rate with which we're dialing and accepting new connections. Maybe `connections` is not the right name then?

If I understand correctly, what you're proposing is an active connection counter. I think I can introduce some counter variables into the `tracer`, then we don't have to worry about the metrics collector restarting.

Does that make sense to you?

Reply:
If you need a gauge, a better way of doing that could be keeping track of opened connects and of closed connections (with the error code) separately. That allows you to implement a gauge as one of multiple possible downstream reports.